daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/PartialNextElementResolver.scala

```diff
@@ -0,0 +1,310 @@
+ * Note that not all InfosetInputter's have a concept of namespaces (json is a
+ * prime example of this). In order to support this, each nextElement method
+ * has a variable called hasNamespace, which is used to specify whether or not
+ * the InfosetInputter supports namespaces. If it does not support namespaces,
+ * then the namespace parameter should be ignored. This may affect how the next
+ * element is determined, and may result in an UnparseError in some cases.
+ */
+sealed trait PartialNextElementResolver extends Serializable {
+
+  /**
+   * Returns One(ElementRuntimeData) if the arguments are associated with
+   * such definition.
+   *
+   * Returns Nope if there is no such mapping.
+   */
+  def maybeNextElement(name: String, nameSpace: String, hasNamespace: Boolean): Maybe[ElementRuntimeData]
+
+  // TODO: PERFORMANCE: We really should be interning all QNames so that comparison of QNames can be pointer equality
+  // or nearly so. We're going to do tons of lookups in hash tables, which will compute the hash code, find it is equal,
+  // compare the entire namespace string character by character, only to say yes, and yes will be by far the vast
+  // bulk of the lookup results.  Pointer equality would be so much faster....
+  //
+  def maybeNextElement(nqn: NamedQName, hasNamespace: Boolean): Maybe[ElementRuntimeData] =
+    maybeNextElement(nqn.local, if (hasNamespace) nqn.namespace.toStringOrNullIfNoNS else null, hasNamespace)
+
+  def currentPossibleNextElements: Seq[ElementRuntimeData]
+
+  final def currentPossibleNamedQNames = currentPossibleNextElements.map { _.namedQName }
+}
+
+/**
+ * This resolver exists to allow the ERD of a quasi-element to have a resolver, but
+ * it should never actually be used, so this just fails if it is.
+ */
+class DoNotUseThisResolver(trd: TermRuntimeData) extends PartialNextElementResolver {
+
+  override def maybeNextElement(local: String, namespace: String, hasNamespace: Boolean): Maybe[ElementRuntimeData] =
+    Assert.invariantFailed("This resolver should never be used.")
+
+  override def toString() = "DoNotUseThisResolver"
+
+  override val currentPossibleNextElements = Seq()
+}
+
+class NoNextElement(trd: TermRuntimeData) extends PartialNextElementResolver {
+
+  override def maybeNextElement(local: String, namespace: String, hasNamespace: Boolean): Maybe[ElementRuntimeData] = {
+    One(new UnexpectedElementErrorERD(Some(trd), local, namespace, currentPossibleNamedQNames))
+  }
+
+  override def toString() = "NoNextElement"
+
+  override val currentPossibleNextElements = Seq()
+
+}
+
+class OnlyOnePossibilityForNextElement(
+  trd: TermRuntimeData,
+  val nextERD: ElementRuntimeData,
+  isRequiredStreamingUnparserEvent: Boolean)
+  extends PartialNextElementResolver {
+
+  val nqn = nextERD.namedQName
+
+  override def maybeNextElement(local: String, namespace: String, hasNamespace: Boolean): Maybe[ElementRuntimeData] = {
+    lazy val sqn = StepQName(None, local, NS(namespace))
```

Comment:
Move sqn into if block where it's used? No need to evaluate if it's never needed.